app/models/puppetclass_lookup_key.rb

```diff
@@ -0,0 +1,35 @@
+class PuppetclassLookupKey < LookupKey
+  has_many :environment_classes, :dependent => :destroy
+  has_many :environments, :through => :environment_classes, :uniq => true
+  has_many :param_classes, :through => :environment_classes, :source => :puppetclass
+  belongs_to :puppetclass, :inverse_of => :lookup_keys
+
+  def param_class
+    param_classes.first
+  end
+
+  def audit_class
+    param_class
+  end
+
+  scoped_search :in => :param_classes, :on => :name, :rename => :puppetclass, :complete_value => true
+
+  scope :smart_class_parameters_for_class, lambda {|puppetclass_ids, environment_id|
+                                             joins(:environment_classes).where(:environment_classes => {:puppetclass_id => puppetclass_ids, :environment_id => environment_id})
+                                           }
+
+  scope :parameters_for_class, lambda {|puppetclass_ids, environment_id|
+                                 override.smart_class_parameters_for_class(puppetclass_ids,environment_id)
+                               }
+
+  scope :smart_class_parameters, lambda { where(:is_param => true).joins(:environment_classes).readonly(false) }
+
+  # def is_smart_class_parameter?
```

Comment:
I guess this method can go away, unless you want to deprecate it and add a compatibility layer. I'm not aware of any plugins that would use it so I'd say remove it.